jane/doc/parallelism-tutorial.md

```diff
@@ -1,38 +1,162 @@
+are no data races. As a consequence, in a program with data races, it's
+difficult if not impossible to reason about the order in which things happen.
+
+As an example, imagine we have two global variables:
+
+```ocaml
+let price_of_gold = ref 0.0
+let initialised = ref false
+```
+
+Now consider two pieces of code running in parallel. One is some startup code
+we'll call domain A:
+
+```ocaml
+(* ... *)
+price_of_gold := calculate_price_of_gold ();
+initialised := true
+(* ... *)
+```
+
+The other, domain B, accesses the variables:
+
+```ocaml
+(* ... *)
+if !initialised then
+  if !price_of_gold < really_good_price_for_gold then
+    buy_lots_of_gold ()
+(* ... *)
+```
+
+What happens? Well, since they're running in parallel, we don't know exactly.
+There is of course an intuitive way to reason it out: at any given point in
+time, each domain will be partway done running, we just never know exactly which
+domain gets to go next. So maybe domain A sets `price_of_gold`, and then domain
+B reads `initialised` and sees that it's `false`, and then domain A sets
+`initialised` to `true`. Or maybe domain A sets both variables before domain B
+checks `initialised`, and then if gold is cheap, domain B heads to the market.
+
+The one thing we want _not_ to happen is for domain B to see that
+`!initialised` is `true` and yet find that `price_of_gold` is still stuck at
+its initial value of `0.0`, thus causing us to `buy_lots_of_gold ()` no matter
+the price. Our intuitive method reassures us that this cannot happen: domain A
+doesn't set `initialised` to `true` until it's already done setting
+`price_of_gold` to something sensible.
+
+Unfortunately, our intuitive reasoning is wrong: this code very much can end up
+calling `buy_lots_of_gold ()` despite `price_of_gold` being high. Since the
+assignments to `price_of_gold` and `initialised` happen in the same domain and
+neither is a fancy [atomic] operation, they can be reordered so that
+`initialised` is set first, _while `price_of_gold` is still zero,_ leading
+domain B to make an expensive mistake.
+
+[atomic]: #atomics
+
+It turns out our intuition assumes what the literature calls
+_sequential consistency:_ essentially, the principle that things happen in order
+they're written, and parallelism just means we're not sure which domain will do
+its next thing. This principle is so essential that it's hard to imagine
+reasoning about a multicore program without it.
+
+Of course, what's happened here is two data races: each of `price_of_gold` and
+`initialised` is being modified by domain A and read by domain B in parallel.
+So the moral of the story is that data races violate sequential consistency.
+
+The good news is that the inverse is also true: any program without data races
+is sequentially consistent. Data-race freedom gives us back the power to reason
+intuitively about our code, no matter how buggy it might get. In our example,
```

Comment:
```suggestion
is sequentially consistent. *Data-race freedom gives us back the power to reason
intuitively about our code, no matter how buggy it might get.* In our example,
```